src/Hamiltonians/HubbardRealSpace.jl

```diff
@@ -20,58 +20,173 @@ local_interaction(f::FermiFS, _) = 0
 function local_interaction(a::SingleComponentFockAddress, b::SingleComponentFockAddress, u)
     return u * dot(occupied_modes(a), occupied_modes(b))
 end
-function local_interaction(fs::CompositeFS, u)
-    return _interactions(fs.components, u)
-end
 
 """
-    _interaction_col(a, bs::Tuple, us::Tuple)
+    nearest_neighbour_interaction(f::SingleComponentFockAddress, σ, geometry::CubicGrid, map::ModeMap)
+    nearest_neighbour_interaction(f1::SingleComponentFockAddress, f2::SingleComponentFockAddress, 
+        Δ, geometry::CubicGrid, map1::ModeMap)
+
+Calculate the nearest neighbour interaction ``\\σ \\sum_{⟨i,j⟩} n_i n_j`` for a single component 
+Fock state `f` or ``Δ \\sum_{⟨i,j⟩} n_{↑,i} n_{↓,j}`` between two Fock states.For a
+multi-component Fock state, return the eigenvalue of
+
+```math
+\\frac{1}{2}\\sum_{⟨i,j⟩, σ, τ} Δ_{σ,τ} a^†_{σ,i}a^†_{τ,j}a_{τ,j}a_{σ,i} ,
+```
 
-Sum the local interactions of the Fock state `a` with all states in `bs` using the
-interaction constants in `us`. This is used to compute all interactions in the column
-below the diagonal of the interaction matrix.
+where `Δ::SMatrix` is a symmetric matrix of interaction constants, `i, j` are the mode indices,
+and `σ`, `τ` are component indices.
+
+See also [`BoseFS`](@ref), [`FermiFS`](@ref), [`CompositeFS`](@ref).
 """
-@inline _interaction_col(a, ::Tuple{}, ::Tuple{}) = 0
-@inline function _interaction_col(a, (b, bs...), (u, us...))
-    return local_interaction(a, b, u) + _interaction_col(a, bs, us)
+function nearest_neighbour_interaction(f::SingleComponentFockAddress, Δ, 
+    geometry::CubicGrid{D,S,B}, map::ModeMap) where {D,S,B}
+    N = length(map)
+    ONR = onr(f)
+    ext_result = 0
+    for i in 1:N
+        for j in 1:D
+            occ_mode = map[i].mode
+            if !B[j] && check_boundary(occ_mode, j, S)
+                continue # skip the dimension if it is hard wall
+            end
+            neigh = neighbor_site(geometry, occ_mode, j)
+            ext_result += ONR[occ_mode] * ONR[neigh]
+        end
+    end
+    return Δ*ext_result
+end
+function nearest_neighbour_interaction(f1::SingleComponentFockAddress, f2::SingleComponentFockAddress, 
+    Δ, geometry::CubicGrid{D,S,B}, map1::ModeMap
+    ) where {D,S,B}
```

Comment:
```suggestion
    Δ, geometry::CubicGrid{D,S}, map1::ModeMap
    ) where {D,S}
```
Only capture type params you're using in the code.